File: daily/2025/2197_Replace_Non_Coprime_Numbers_in_Array.cpp
#include <iostream>
#include <vector>
#include <stack>

using namespace std;

class Solution {
public:
    long long gcd(long long a, long long b){
        if (b == 0) return a;
        return gcd(b, a % b);
    }
    vector<int> replaceNonCoprimes(vector<int>& nums) {
        stack<long long> s;
        for (int curr : nums){
            while(!s.empty() && gcd(curr, s.top()) > 1){
                curr = curr / gcd(curr, s.top()) * s.top();
                s.pop();
            }
            s.push(curr);
        }
        vector<int> ans;
        while(!s.empty()){
            ans.push_back(s.top());
            s.pop();
        }
        reverse(ans.begin(), ans.end());
        return ans;
    }
};

int main(){
    vector<int> nums = {6,4,3,2,7,6,2};
    // vector<int> nums = {2,2,1,1,3,3,3};
    // vector<int> nums = {287,41,49,287,899,23,23,20677,5,825};
    Solution S;
    vector<int> ans = S.replaceNonCoprimes(nums);
    for (int i : ans) cout << i << " ";
    cout << endl;
    return 0;
}

File: daily_python/2197_Replace_Non_Coprime_Numbers_in_Array.py
import os
import sys
from typing import List

class Solution:
    def replaceNonCoprimes(self, nums: List[int]) -> List[int]:
        def gcd(x, y):
            if y == 0:
                return x
            return gcd(y, x % y)
        
        s = []
        for curr in nums:
            while s and gcd(curr, s[-1]) > 1:
                curr = curr / gcd(curr, s[-1]) * s[-1]
                s.pop()
            s.append(int(curr))
        return s

def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            nums = list(map(int, lines[0].split('=')[1].strip()[1:-2].split(',')))
            dataset.append(nums)
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for nums in dataset:
        results.append(solution.replaceNonCoprimes(nums))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')